### PR TITLE
[Auto Parallel] fix save load state_dict

### PR DESCRIPTION
**Pull request by xingmingyyj**
### PR Category

Auto Parallel 

### PR Types

Others

### Description

fix save load state_dict

### CODE REVIEW

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 1]
```diff
@@ -437,6 +439,15 @@ def load_state_dict(
         rank_to_files, missing_keys = get_rank_to_files(
             path, flat_state_dict, process_group, use_dist
         )
+        
+        gloabl_rank_to_files = []
```
Comment (zhangbo9674) [inline python/paddle/distributed/checkpoint/load_state_dict.py]: get_rank_to_files 已经是通过 global_data_files 分析的结果，这里还需要 all_gather_object(gloabl_rank_to_files)么？

Comment (xingmingyyj) [reply python/paddle/distributed/checkpoint/load_state_dict.py]: 这里不需要all_gather,但是get_rank_to_files中要做一些修改。动半和静半得到的state_dict不同，在静半下，state_dict是部分的，所以的到的necessary_files也是部分的，所以需要对necessary_files做一下all_gather。

**Review by zhangbo9674 — APPROVED** [review]
LGTM

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 2]
```diff
@@ -549,9 +566,11 @@ def load_state_dict(
                         storage_chunk_tensor, src=src_rank, group=process_group
                     )
                 else:
+                    tmp_tensor = paddle.assign(cur_chunk_tensor)
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/load_state_dict.py:569]: why use tmp_tensor, plz add comments.

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 3]
```diff
@@ -533,14 +551,13 @@ def load_state_dict(
...
-                    dtype=flat_state_dict[
-                        item.local_tensor_index.tensor_key
-                    ].dtype,
+                    item.dtype,
                 )
 
             if src_rank == item.rank:
-                # assign value locally
-                paddle.assign(storage_chunk_tensor, cur_chunk_tensor)
+                if src_rank == paddle.distributed.get_rank():
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/load_state_dict.py:558]: what about the else branch?

File: python/paddle/distributed/checkpoint/metadata.py [hunk 4]
```diff
@@ -24,6 +24,7 @@ class LocalTensorMetadata:
 
     global_offset: Tuple[int]
     local_shape: Tuple[int]
+    dtype: str
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/metadata.py:27]: why we need dtype?

File: python/paddle/distributed/checkpoint/load_state_dict.py (hunk 2 above)
Comment (xingmingyyj) [reply python/paddle/distributed/checkpoint/load_state_dict.py:569]: The memory held by cur_chunk_tensor may be non-contiguous, and the broadcast API does not support this type of tensor.

File: python/paddle/distributed/checkpoint/load_state_dict.py (hunk 3 above)
Comment (xingmingyyj) [reply python/paddle/distributed/checkpoint/load_state_dict.py:558]: The condition ``src_rank == item.rank `` will be satisfied by all ranks, but only one rank needs to perform the assignment operation. Additionally, when ``src_rank != paddle.distributed.get_rank()``, ``storage_chunk_tensor`` may be None.

File: python/paddle/distributed/checkpoint/metadata.py (hunk 4 above)
Comment (xingmingyyj) [reply python/paddle/distributed/checkpoint/metadata.py:27]: In static mode, the ``state_dict`` is incomplete, so the previous code would trigger  ``key error``. Here, the ``dtype`` is stored in advance.

File: python/paddle/distributed/checkpoint/load_state_dict.py (hunk 3 above)
Comment (xingmingyyj) [reply python/paddle/distributed/checkpoint/load_state_dict.py:558]: We’ll add comments in the next PR.

**Pull request merged**